Improve speed of GROMOS11 reader

schuhmc opened the pull request:
<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #5079

The GROMOS11 trajectory reader written by @JoStoe was merged in #4292 (see issue #4291 )

The performance in the current state is very slow

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 

- Use seek to skip over the POSITIONRED block when creating the Universe
- Prevent calling strip() on the same line multiple times
- Improved reading speed of the POSITIONRED block by using a buffer and processing with numpy.fromstring()
- Pre-allocate numpy arrays instead of reading into list() and calling numpy.asarray()
- Add the GROMOS11 reader to the benchmark set
- Add a test with a trajectory containing inconsistent whitespace

All-in-all this provides a significant speed boost. Internal testing showed the following speed increase when loading and iterating over a 1 GB trajectory of 2000 frames:

Before these changes: 86.97 s +/- 1.13 s (n=10)
With changes in this PR: 42.05 s +/- 0.66 s (n=10)
Speed improvement -51.6%

On the newly added benchmark the following speed improvement was measured:
traj_reader.TrajReaderCreation.time_reads: 5.339 ms -> 3.579 ms (-33.0%)
traj_reader.TrajReaderIteration.time_strides: 7.642 ms -> 4.524 ms (-40.8%)

(Note that the benchmark trajectories are very small which reduces the impact on the TrajReaderCreation.time_reads benchmark. Based on testing on larger trajectories, the time for the Universe creation is reduced by 70-80% if the file has consistent block sizes)

Most of the speed improvement comes from the skipping over the POSITIONRED block, meaning that it is not necessary to read the full trajectory upon universe creation. 
However, note that when the size of the POSITIONRED block is not consistent (whitespace), the speed is impacted.

## PR Checklist
 - [x] Issue raised/referenced?
 - [x] Tests updated/added?
 - [ ] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

I don't think that the documentation needs to be updated.

## Developers Certificate of Origin

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5080.org.readthedocs.build/en/5080/

<!-- readthedocs-preview mdanalysis end -->

schuhmc committed:
skip conversion from list to np.array

schuhmc committed:
strip line only once

schuhmc committed:
Implemented seeking the POSITIONRED block

schuhmc committed:
Further performance improvements

schuhmc committed:
performance improvements

schuhmc committed:
add benchmark for trc files

schuhmc committed:
add AUTHORS entry. Add Tests

schuhmc committed:
update CHANGELOG

schuhmc committed:
run black; change TRC file for benchmark

orbeckst commented on benchmarks/benchmarks/traj_reader.py:
You're AWESOME! It's great that you added benchmark!!

schuhmc replied:
Thanks! Note that I had some troubles getting the benchmark suite running without downgrading asv to 0.6.1 (seems to be related to https://github.com/airspeed-velocity/asv/issues/1453 ). I will try to reproduce and create an issue later today.

orbeckst commented on package/MDAnalysis/coordinates/TRC.py:
```diff
@@ -194,60 +194,95 @@ def _read_traj_properties(self):
-
-                if first_block is not None:
-                    break  # First block found
+                stripped_line = line.strip()
```
You could just re-assign to `line` 
```
line = line.strip()
```
if you want to. I think it makes the following code slightly less noisy but you decide what you prefer.

orbeckst replied:
This comment goes for all occurences of the same construct in your code.

schuhmc replied:
In the loop at line 218 I need to get the correct length of the non-stripped line to seek to the correct position in the file afterwards. If I re-assign `line = line.strip()`, the position will be off by a non-predictable amount.
I could do the re-assignment in this particular loop and also in the _read_GROMOS11_trajectory loop, but I prefer to keep it consistent across the different loops.

orbeckst replied:
Good explanation, thanks, maybe worthwhile adding as a comment to the code, to avoid overeager optimization by others in the future.

orbeckst commented on package/MDAnalysis/coordinates/TRC.py:
```diff
@@ -194,60 +194,95 @@ def _read_traj_properties(self):
+                    if n_atoms == 0:
+                        start = f.tell() - len(line)
+                        while stripped_line != "END":
+                            line = f.readline()
+                            stripped_line = line.strip()
```
merge as in other parts of the code?
```
stripped_line = f.readline().strip()
```

schuhmc replied:
I do actually need the length of the non-stripped line when exiting the loop. Merging would mean I loose this information.

orbeckst commented on package/MDAnalysis/coordinates/TRC.py:
```diff
@@ -194,60 +194,95 @@ def _read_traj_properties(self):
+                        # to where we were before
+                        if f.readline().strip() != "END":
+                            inconsistent_size = True
+                            warnings.warn(
```
Also use `logger.warn()`, given that we're already using the logger. 

(We haven't been using logging consistently but we should because otherwise it's difficult to get proper diagnostics in complicated workflows.)

orbeckst replied:
I'd do something like
```python
warnmsg = "Inconsistent POSITIONRED block size. Falling back to slow reader."
warnins.warn(warnmsg, UserWarning)
logger.warn(warnmsg)
```
If you can embed the filename in the warnmsg then it may be even more helpful.

schuhmc replied:
the warn() method seems to be deprecated. (see https://docs.python.org/3/library/logging.html#logging.Logger.warning ). I will change it to warning() as suggested by the documentation.

orbeckst commented on package/MDAnalysis/coordinates/TRC.py:
```diff
@@ -274,69 +309,81 @@ def _read_GROMOS11_trajectory(self):
+                if i != self.traj_properties["n_atoms"]:
                     raise ValueError(
-                        "The trajectory contains the wrong number of atoms!"
+                        "The trajectory contains the too few atoms!"
```
Optional: 
* State found `i` and expected n_atoms in errmsg.
* Also use `logger.error()`.

orbeckst reviewed (approved):
Looks really good! Thank you for your contribution. Great work.

I only have a few optional comments. If you want to address them, please let us know how much time you need. If you don't want to address them please say so, too, and then we can merge right away.

EDIT: Please run `black` over TRC.py and check all the other files that you touched. We won't merge until the linter is also happy.

schuhmc committed:
add PR number to changelog

schuhmc committed:
changed warning for inconsistent size. Adjusted test.

schuhmc committed:
increase verbosity of error message for position count mismathc

schuhmc committed:
remove debug message

schuhmc committed:
unify errors and warnings. Add test for extra/missing positions in traj

schuhmc committed:
run black

schuhmc commented:
Thank you for the valuable feedback @orbeckst !

I implemented the requested changes where possible (see comments). The linters also seem happy now.
I also added tests for trajectories with too many/few positions.

From my side this should be ready to merge.

orbeckst commented:
Thank you @schuhmc , very nice contribution! 🎉
